File: 2015/rust/src/day5.rs
use std::collections::HashMap;
use std::error::Error;
use std::fs::File;
use std::io::prelude::*;

pub fn day5(input_file: &str) -> Result<(), Box<dyn Error>> {
    println!("Day 5: Doesn't He Have Intern-Elves For This?");

    let mut f = File::open(input_file)?;
    let mut input = String::new();
    f.read_to_string(&mut input)?;

    println!("\tPart 1: {}", part1(&input)?);
    println!("\tPart 2: {}", part2(&input)?);
    Ok(())
}

fn part1(input: &str) -> Result<i32, Box<dyn Error>> {
    let mut nice = 0;
    'outer: for line in input.lines() {
        let mut vowels = 0;
        let mut last = '\0';
        let mut double = false;

        for c in line.chars() {
            match c {
                'b' => {
                    if last == 'a' {
                        continue 'outer;
                    }
                }
                'd' => {
                    if last == 'c' {
                        continue 'outer;
                    }
                }
                'q' => {
                    if last == 'p' {
                        continue 'outer;
                    }
                }
                'y' => {
                    if last == 'x' {
                        continue 'outer;
                    }
                }
                _ => (),
            };
            match c {
                'a' | 'e' | 'i' | 'o' | 'u' => vowels += 1,
                _ => (),
            };
            if c == last {
                double = true;
            }
            last = c;
        }
        if double == true && vowels >= 3 {
            nice += 1;
        }
    }
    Ok(nice)
}

fn part2(input: &str) -> Result<i32, Box<dyn Error>> {
    let mut nice = 0;
    'outer: for line in input.lines() {
        let mut pairs: HashMap<String, Vec<usize>> = HashMap::new();
        let mut last = '\0';
        let mut two_ago = '\0';
        let (mut split, mut two_pair) = (false, false);

        for (i, c) in line.chars().enumerate() {
            let mut pair = String::new();
            pair.push(last);
            pair.push(c);
            let locs = pairs.entry(pair).or_insert(Vec::new());
            locs.push(i);

            if c == two_ago {
                split = true;
            }
            let locslen = locs.len();
            if (locslen > 1 && (locs[locslen - 1] - locs[locslen - 2] > 1)) || locslen > 2 {
                two_pair = true;
            }
            if split && two_pair {
                nice += 1;
                continue 'outer;
            }
            two_ago = last;
            last = c;
        }
    }
    Ok(nice)
}

File: 2015/c/src/day5.c
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <pcre.h>
#include "lib.h"

static int part1(const char *input, size_t input_len)
{
    const char *error;
    int erroffset;
    int good = 0;

    pcre *nice_re = pcre_compile("^(?!.*(?:ab|cd|pq|xy))(?=.*(.)\\1)(?=(?:.*[aeiou]){3,}).*$", 0, &error, &erroffset, NULL);
    if (nice_re == NULL)
    {
        errno = ENOTRECOVERABLE;
        goto err_cleanup;
    }

    char **lines;
    size_t lines_len = input_to_lines(&lines, input, input_len);
    if (lines_len < 0)
        goto err_cleanup;

    for (int i = 0; i < lines_len; i++)
    {
        if (!pcre_exec(nice_re, NULL, lines[i], strlen(lines[i]), 0, 0, NULL, 0))
            good++;
    }

    goto cleanup;

err_cleanup:
    good = -1;
cleanup:
    if (nice_re != NULL)
        pcre_free(nice_re);
    if (lines != NULL)
        free_lines(lines, lines_len);
    return good;
}

static int part2(const char *input, size_t input_len)
{
    const char *error;
    int erroffset;
    int good = 0;

    pcre *nice_re = pcre_compile("^(?=.*(..).*\\1)(?=.*(.).\\2).*$", 0, &error, &erroffset, NULL);
    if (nice_re == NULL)
    {
        errno = ENOTRECOVERABLE;
        goto err_cleanup;
    }

    char **lines;
    size_t lines_len = input_to_lines(&lines, input, input_len);
    if (lines_len < 0)
        goto err_cleanup;

    for (int i = 0; i < lines_len; i++)
    {
        if (!pcre_exec(nice_re, NULL, lines[i], strlen(lines[i]), 0, 0, NULL, 0))
            good++;
    }

    goto cleanup;

err_cleanup:
    good = -1;
cleanup:
    if (nice_re != NULL)
        pcre_free(nice_re);
    if (lines != NULL)
        free_lines(lines, lines_len);
    return good;
}

int day5(const char *filename)
{
    printf("Day 5: Doesn't He Have Intern-Elves For This?\n");

    char *input;
    int filesize = read_file_to_buffer(&input, filename);
    if (filesize < 0)
        return -1;

    int rval = part1(input, filesize);
    if (rval < 0)
        return rval;
    printf("\tPart 1: %d\n", rval);

    rval = part2(input, filesize);
    if (rval < 0)
        return rval;
    printf("\tPart 2: %d\n", rval);

    free(input);

    return 0;
}
